x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameIndexer.java

```diff
@@ -61,24 +78,32 @@
 
     private Pivot pivot;
     private int pageSize = 0;
-    protected volatile DataFrameTransformCheckpoint inProgressOrLastCheckpoint;
+    protected volatile DataFrameTransformCheckpoint lastCheckpoint;
+    protected volatile DataFrameTransformCheckpoint nextCheckpoint;
+
+    private volatile RunState runState;
+
+    // hold information for continuous mode (partial updates)
     private volatile Map<String, Set<String>> changedBuckets;
+    private volatile Map<String, Object> changedBucketsAfterKey;
 
     public DataFrameIndexer(Executor executor,
                             DataFrameAuditor auditor,
                             DataFrameTransformConfig transformConfig,
                             Map<String, String> fieldMappings,
                             AtomicReference<IndexerState> initialState,
-                            Map<String, Object> initialPosition,
+                            DataFrameIndexerPosition initialPosition,
                             DataFrameIndexerTransformStats jobStats,
                             DataFrameTransformProgress transformProgress,
-                            DataFrameTransformCheckpoint inProgressOrLastCheckpoint) {
+                            DataFrameTransformCheckpoint lastCheckpoint,
+                            DataFrameTransformCheckpoint nextCheckpoint) {
         super(executor, initialState, initialPosition, jobStats);
         this.auditor = Objects.requireNonNull(auditor);
         this.transformConfig = ExceptionsHelper.requireNonNull(transformConfig, "transformConfig");
         this.fieldMappings = ExceptionsHelper.requireNonNull(fieldMappings, "fieldMappings");
         this.progress = transformProgress;
-        this.inProgressOrLastCheckpoint = inProgressOrLastCheckpoint;
+        this.lastCheckpoint = lastCheckpoint;
+        this.nextCheckpoint = nextCheckpoint;
     }
```

Comment:
The constructor doesn't set `runState`, so it starts off `null`.  This will cause an NPE if any `switch (runState)` is called while it's `null`.

Would it be appropriate to call `runState = determineRunStateAtStart();` in the constructor?  Or if that would cause problems then consider changing:

```
    switch (runState) {
        case x:
            return x();
        case y:
            return y();
        case z:
            return z();
        default:
            logger.warn("wrong");
            throw error;
    }
```

to:

```
    if (runState != null) {
        switch (runState) {
            case x:
                return x();
            case y:
                return y();
            case z:
                return z();
        }
    }
    logger.warn("wrong");
    throw error;
```

It would give us more information if `runState` still was `null` at the time of a `switch`.

Reply:
`onStart` gets called before any of the reads of `runState` happen, so an NPE should not be possible. 

Because `determineRunStateAtStart()` depends on steps before, it would not return the same value if moved to the constructor.

However, I will assign a default to runState at the constructor, to be on the safe side. 

Good spot!